File: public/js/record.js
// A record's identity on the chain: the hash of its canonical payload and its
// signature together. Mirrors lib/reputable_chat/cryptography/record.rb --
// spec/record_parity_spec.rb checks that they agree.

import * as canonical from "./canonical.js";

export const DOMAIN = "reputablechat:record:v1";
const SEPARATOR = "\n";
const HEX = /^[0-9a-f]{64}$/;

// Takes the canonical payload as a string wherever one is already to hand, so
// a blob the server handed back is hashed as the bytes it actually sent rather
// than as a re-serialization of them.
export async function digest(payload, signature) {
  const body = typeof payload === "string" ? payload : canonical.dump(payload);

  // The separators are unambiguous rather than conventional: canonical JSON
  // can never hold a raw newline, and base64url has none either, so exactly
  // one pair of strings produces any given hash input. Checked because the
  // claim is load-bearing.
  if (body.includes(SEPARATOR)) throw new Error("canonical payload contains a newline");
  if (String(signature).includes(SEPARATOR)) throw new Error("signature contains a newline");

  const input = new TextEncoder().encode([DOMAIN, body, signature].join(SEPARATOR));
  const hashed = await crypto.subtle.digest("SHA-256", input);

  return [...new Uint8Array(hashed)].map((b) => b.toString(16).padStart(2, "0")).join("");
}

export function isValid(value) {
  return typeof value === "string" && HEX.test(value);
}

File: lib/reputable_chat/cryptography/record.rb
# frozen_string_literal: true

require "digest"
require_relative "canonical"

module ReputableChat
  module Cryptography
    # A record's identity on the chain: the hash of its canonical payload and
    # its signature together.
    #
    # `ack`, `reply_to` and an emote's `message` field all name one of these.
    # Never a signature: a signature identifies a payload, while a record hash
    # identifies the whole record, signature included, which is what a link has
    # to cover if it is going to be tamper-evident.
    #
    # Must agree with public/js/record.js -- spec/record_parity_spec.rb is what
    # catches it if it ever does not.
    module Record
      DOMAIN    = "reputablechat:record:v1"
      SEPARATOR = "\n"
      HEX       = /\A[0-9a-f]{64}\z/

      class MalformedPayload < StandardError; end

      module_function

      # Takes the canonical payload as a String wherever one is already to
      # hand. The server stores the bytes exactly as they arrived and hashes
      # those; re-serializing a parsed payload server-side is the one habit
      # guaranteed to break a signature eventually.
      def digest(payload:, signature:)
        canonical = payload.is_a?(String) ? payload : Canonical.dump(payload)

        # The separators are unambiguous rather than merely conventional:
        # canonical JSON can never hold a raw newline (JSON escapes one inside
        # a string to the two characters \n, and there is no whitespace between
        # tokens) and base64url has none either. So exactly one pair of strings
        # produces any given hash input. Checked rather than assumed, because
        # the claim is load-bearing and costs one scan to keep honest.
        raise MalformedPayload, "canonical payload contains a newline" if canonical.include?(SEPARATOR)
        raise MalformedPayload, "signature contains a newline" if signature.to_s.include?(SEPARATOR)

        Digest::SHA256.hexdigest(
          [DOMAIN, canonical, signature].join(SEPARATOR).b
        )
      end

      def valid?(value) = value.is_a?(String) && value.match?(HEX)
    end
  end
end
